src/modules/0.0.21/aws_codebuild/mappers/project.ts

```diff
@@ -0,0 +1,173 @@
+    res => res?.project,
+  );
+
+  getProjects = crudBuilderFormat<CodeBuild, 'batchGetProjects', Project[] | undefined>(
+    'batchGetProjects',
+    input => input,
+    res => res?.projects,
+  );
+
+  listProjects = paginateBuilder<CodeBuild>(paginateListProjects, 'projects');
+
+  deleteProject = crudBuilder2<CodeBuild, 'deleteProject'>('deleteProject', input => input);
+
+  cloud: Crud2<CodebuildProject> = new Crud2({
+    create: async (es: CodebuildProject[], ctx: Context) => {
+      const client = (await ctx.getAwsClient()) as AWS;
+      const out = [];
+      for (const e of es) {
+        const input: CreateProjectCommandInput = {
+          name: e.projectName,
+          environment: {
+            type: e.environmentType,
+            image: e.image,
+            computeType: e.computeType,
+            environmentVariables: e.environmentVariables,
+            privilegedMode: e.privilegedMode,
+          },
+          sourceVersion: e.sourceVersion,
+          source: {
+            location: e.sourceLocation,
+            type: e.sourceType,
+            buildspec: e.buildSpec,
+          },
+          serviceRole: e.serviceRole?.arn,
+          // TODO implement artifacts
+          artifacts: {
+            type: 'NO_ARTIFACTS',
+          },
+        };
+        const awsPj = await this.createProject(client.cbClient, input);
+        if (!awsPj) continue;
+        const newPj = await this.projectMapper(awsPj, ctx);
+        if (newPj) out.push(newPj);
+      }
+      return out;
+    },
+    read: async (ctx: Context, id?: string) => {
+      const client = (await ctx.getAwsClient()) as AWS;
+      if (id) {
+        const input: BatchGetProjectsCommandInput = {
+          names: [id],
+        };
+        const pjs = await this.getProjects(client.cbClient, input);
+        if (!pjs || pjs.length !== 1) return;
+        const pj = await this.projectMapper(pjs[0], ctx);
+        if (!pj) return;
+        return pj;
+      } else {
+        const pjIds = await this.listProjects(client.cbClient);
+        if (!pjIds || !pjIds.length) return;
+        const input: BatchGetProjectsCommandInput = {
+          names: pjIds,
+        };
+        const pjs = await this.getProjects(client.cbClient, input);
+        if (!pjs) return;
+        const out = [];
+        for (const pj of pjs) {
+          const outPj = await this.projectMapper(pj, ctx);
+          if (outPj) out.push(outPj);
+        }
+        return out;
+      }
+    },
+    updateOrReplace: (_a: CodebuildProject, _b: CodebuildProject) => 'replace',
+    update: async (pjs: CodebuildProject[], ctx: Context) => {
+      const out = [];
+      for (const pj of pjs) {
+        const cloudRecord = ctx?.memo?.cloud?.CodebuildProject?.[pj.projectName ?? ''];
+        if (pj.arn !== cloudRecord.arn) {
+          pj.arn = cloudRecord.arn;
+          if (this.module.project.equals(pj, cloudRecord)) {
+            await this.module.project.db.update(cloudRecord, ctx);
```

Comment:
nit: if we are restoring the record here technically it is not a `replace` but an `update` and the `updateOrReplace` should return the correct string?

Reply:
I also copied this over from the ACM module

Reply:
I thought the `updateOrReplace` was only called explicitly in the mapper and not in the interface?

Reply:
hmm not sure what do you mean? What I wanted to say is that the `updateOrReplace` method for this mapper is returning always `replace` and in the `update` method we are doing a restore, so that could lead to confusion in the code and to the user when we show the changes made or to be applied.